### PR TITLE
OpenAI Vision image query functionality

### PR DESCRIPTION
**Pull request by andrewfrench**
This PR introduces image query functionality using the OpenAI Vision model.

### CODE REVIEW

File: CHANGELOG.md
Comment (collindutter) [inline CHANGELOG.md:17]: Capitalize `engine` and `task`

File: griptape/drivers/image_query/base_image_query_driver.py
Comment (collindutter) [inline griptape/drivers/image_query/base_image_query_driver.py:30]: Should we refactor to `ListArtifact`? Thinking not now, but just raising it.

File: griptape/drivers/image_query/openai_vision_image_query_driver.py
Comment (collindutter) [inline griptape/drivers/image_query/openai_vision_image_query_driver.py:23]: Optional?

File: griptape/drivers/image_query/openai_vision_image_query_driver.py
Comment (collindutter) [inline griptape/drivers/image_query/openai_vision_image_query_driver.py:56]: Should we add a check similar to the chat prompt driver that verifies that there is exactly one choice?

File: griptape/events/base_task_event.py
Comment (collindutter) [inline griptape/events/base_task_event.py:17]: Is that a duplicate `BaseArtifact`? I know not related to your change, but can we clean that up?

File: griptape/tasks/image_query_task.py
Comment (collindutter) [inline griptape/tasks/image_query_task.py:18]: If you add `alias="input"` users can set the field using `input` instead of `_input`.

File: griptape/tasks/image_query_task.py
Comment (collindutter) [inline griptape/tasks/image_query_task.py:15]: May want to merge in #591 first so that we can do something similar to [this](https://github.com/griptape-ai/griptape/pull/591/files#diff-30ce1ef1ca99f2869a727dd437c0a9985d34cddc1e9a922142bdf4020bd366bfR49)

File: griptape/tasks/image_query_task.py
Comment (collindutter) [inline griptape/tasks/image_query_task.py:33]: Capitalize artifact names?

File: griptape/tasks/image_query_task.py
Comment (collindutter) [inline griptape/tasks/image_query_task.py:43]: I _think_ with the `alias` change we can get rid of the setter.

File: griptape/tasks/image_query_task.py
Comment (collindutter) [inline griptape/tasks/image_query_task.py:47]: Total nit but if we wanna get pythonic with it we can do `query, image_artifacts = self.input`

File: griptape/tools/image_query_client/tool.py
Comment (collindutter) [inline griptape/tools/image_query_client/tool.py:65]: Is this usually how we structure the schema for Task Memory fields? Isn't it typically without descriptions and flat?

File: griptape/drivers/image_query/openai_vision_image_query_driver.py
Comment (emjay07) [inline griptape/drivers/image_query/openai_vision_image_query_driver.py:40]: what is the expected behavior for a list of images? does it string all of the text together? wondering why defaulting to multiple?

File: griptape/drivers/image_query/openai_vision_image_query_driver.py
Comment (emjay07) [inline griptape/drivers/image_query/openai_vision_image_query_driver.py:52]: if the user doesn't provide max_tokens, what does OpenAI do? does it have a default?

File: griptape/tasks/image_query_task.py
Comment (emjay07) [inline griptape/tasks/image_query_task.py:33]: nit: can we add docs to this task explaining the required order params? I saw that the image_eneration tasks had helpful class-level docs.

File: griptape/tools/image_query_client/tool.py
Comment (emjay07) [inline griptape/tools/image_query_client/tool.py:35]: not sure if this is a pattern we've used yet: but should this be using the file_manager tool? it has a whole advanced interface for picking the proper loader and error checking for it.

File: griptape/tasks/image_query_task.py
Comment (andrewfrench) [reply griptape/tasks/image_query_task.py:43]: I think this just allows us to reference it by the alias in the constructor, like `ImageQueryTask(input=...`

>    :param str | None alias: Override this attribute's parameter name in the
        generated ``__init__`` method. If left `None`, default to ``name``
        stripped of leading underscores. See `private-attributes`.

File: griptape/tasks/image_query_task.py
Comment (andrewfrench) [reply griptape/tasks/image_query_task.py:47]: Clean.

File: griptape/tools/image_query_client/tool.py
Comment (andrewfrench) [reply griptape/tools/image_query_client/tool.py:65]: Good point, I removed the descriptions. I don't think we have other examples that behave quite like this. Here, the agent can specify a series of namespace-artifact combinations, which would be important if multiple images are provided as inputs from independent tasks.

File: griptape/drivers/image_query/openai_vision_image_query_driver.py
Comment (andrewfrench) [reply griptape/drivers/image_query/openai_vision_image_query_driver.py:40]: This enables fundamentally different behavior than operating serially over a list of images. As implemented, we can provide many images to Vision and query it in reference to _all the images at once_, not just one at a time. An example would be providing two images and asking the model "What's the difference between these two images?" to understand that one contains a dog and the other contains a cat.

Of course, we can always provide a list of length one to query a single image.

File: griptape/drivers/image_query/openai_vision_image_query_driver.py
Comment (andrewfrench) [reply griptape/drivers/image_query/openai_vision_image_query_driver.py:52]: Unfortunate backstory: OpenAI's APIs have an outstanding issue where your account's rate limit is penalized for the max tokens requested, not just the tokens returned. This is why the `OpenAiChatPromptDriver` only provides `max_tokens` only when explicitly provided. Vision, specifically, also has another complicating issue where it will default to a very small number of output tokens (like 16) unless otherwise requested. 

The interaction of both of these issues means there really isn't a great way to define a default here. Requesting a high `max_tokens` value will very quickly consume a typical user's rate limit. The right number will depend on the context.

File: griptape/tools/image_query_client/tool.py
Comment (andrewfrench) [reply griptape/tools/image_query_client/tool.py:35]: This pattern is used in text+image to image generation tools like VariationImageGenerationClient, InpaintingImageGenerationClient, and OutpaintingImageGenerationClient to provide some flexibility on the source of the input images. I don't think FileManager would make sense here because we only want to load images, so we don't need that loader selection logic.

File: griptape/tasks/image_query_task.py
Comment (andrewfrench) [reply griptape/tasks/image_query_task.py]: Good call! Added a docstring.

File: griptape/tasks/image_query_task.py
Comment (collindutter) [reply griptape/tasks/image_query_task.py:53]: Reposting conversation from [other PR](https://github.com/griptape-ai/griptape/pull/591/files#r1482277530)

File: griptape/tools/image_query_client/tool.py
Comment (emjay07) [reply griptape/tools/image_query_client/tool.py:35]: That's fair. It's just using the ImageLoader. I was more thinking on the validation side. Are there any validations we need to do here to make sure the user is handing us an image file? does it just blow up if it's not an image file?

File: griptape/tasks/image_query_task.py
Comment (andrewfrench) [reply griptape/tasks/image_query_task.py:53]: Changed to avoid Pyright errors in the future.

**Review by collindutter — DISMISSED** [review]
Overall looks great. Up to you whether you'd like to wait until #591 is merged in first.

**Review by emjay07 — DISMISSED** [review]

File: griptape/tools/image_query_client/tool.py
Comment (emjay07) [reply griptape/tools/image_query_client/tool.py:35]: discussed in person. the underlying parser will properly surface errors if there are problems parsing.

**Comment by collindutter** [comment]
@andrewfrench 591 went in, let me know if you have any questions on how to integrate here.

File: griptape/tasks/image_query_task.py
Comment (collindutter) [inline griptape/tasks/image_query_task.py:64]: Can you please add a unit test for this logic? Should be nearly identical to the unit tests for the other Tasks.

**Review by collindutter — CHANGES_REQUESTED** [review]

File: griptape/config/openai_structure_config.py
Comment (andrewfrench) [reply griptape/config/openai_structure_config.py:1]: Right, Bedrock will inherit the Dummy driver.

**Review by collindutter — APPROVED** [review]

**Pull request merged**
